Writing Flow: 'Down' key at bottom creates new paragraph

mcsf opened the pull request:
## Description
Keyboard-only alternative to #3623 (see #3078).

- When at the bottom of the editor's content, pressing the down arrow `↓` focuses a brand new paragraph (or whichever is the default block type).
- The initial motivation was the fact that, whenever relying on the keyboard for quick editing and insertion of blocks, it becomes cumbersome to manually _break out_ of most blocks and create a new one.
- Blocks like Paragraph, Heading or List allow the user to split into a new empty block by pressing `Enter` at the edge of the block, but this isn't the case for other blocks.
- However, pressing `Enter` at the edge of a Quote block just adds a line within the citation field of that Quote. And while `Enter` does work for e.g. Image, right now this isn't visually clear (see #3951 for improvements).
- We could try to generically pick up these `Enter` events and append an empty block, but maybe there is value in adding this `↓`-based mechanism—regardless of `Enter` improvements.

## How Has This Been Tested?
1. Add a Quote
2. Fill fields _quote_ and _citation_
3. From the _citation_ field, expect to press `Enter` to move on to a new paragraph (or indeed _any_ block thanks to `/` autocompletion, which I use with a passion).
4. Notice how `Enter` adds a line break within _citation_, so press `Backspace` to undo
5. Press `↓`, notice how the caret has moved into a new empty block.

## Screenshots (jpeg or gifs if applicable):
![gutenberg-writing-flow-down-arrow](https://user-images.githubusercontent.com/150562/33943183-46b5229e-e010-11e7-824b-fca8e4310587.gif)

## Types of changes
Writing flow enhancement.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style.
- [ ] My code follows has proper inline documentation.

mcsf committed:
Writing Flow: 'Down' key at bottom creates new paragraph

mcsf committed:
Refactor by adding appendDefaultBlock action

jasmussen commented:
I love it! All of these little flow improvements really add up, and this is a nice one. I still think we need #3078 so you can also _click_ at the end of a block list (after an image block for example), but this feels like it ties into that. 

Looks like there's a visual regression in the blockquote citation. The font should be smaller and gray, which it is in master.

youknowriad commented on editor/components/default-block-appender/index.js:
```diff
@@ -9,27 +9,16 @@ import 'element-closest';
+import { appendDefaultBlock } from '../../actions';
 import { getBlockCount } from '../../selectors';
 
 export class DefaultBlockAppender extends Component {
```
I guess this could be a functional component now :)

mcsf replied:
My thinking too, but jest wasn't happy and I left it for now :)

youknowriad reviewed (commented):
Some questions:

Do you think we should allow this behavior on "text" blocks only (all blocks with caret) or we should use "enter" in these cases?

If we were to do so, why it's not working on empty paragraphs?

mcsf commented:
> Do you think we should allow this behavior on "text" blocks only (all blocks with caret) or we should use "enter" in these cases?

Not sure what you mean. Do you mean should the `↓` append a default block only when the block before it is a textual block? On first thought it seems it should be the behavior for all blocks (and the implementation kind of achieves this by looking at tabbables).

> If we were to do so, why it's not working on empty paragraphs?

Similarly to how #3623 had [this issue](https://github.com/WordPress/gutenberg/pull/3623#issuecomment-346605599) with endless appending, the current implementation bails if the last tabbable is empty. One glaring issue with that is the following, which can be reproduced:
1. Add a Quote block.
2. Fill out _quote_ but not _citation_.
3. Keeping pressing `↓`, hoping to move into an empty appendage, but nothing happens.

Perhaps we can be smarter and look not just at the last tabbable (in this case, the `footer` inside Quote) but also its tabbable siblings within the block. If at least one of those is non empty, then allow the appending and switching into an empty default block.

mcsf committed:
WritingFlow: 'down' appends new block after any non-empty block

Fixes the following issue introduced by the parent commits:

1. Add a Quote block.
2. Fill out quote but not citation.
3. Keeping pressing ↓, hoping to move into an empty appendage, but
   nothing happens.

Also correctly account for block-level-only tabbables, e.g. Image.

mcsf commented:
56e22cd addresses my previous comment:

> Perhaps we can be smarter and look not just at the last tabbable (in this case, the `<footer>` inside Quote) but also its tabbable siblings within the block. If at least one of those is non empty, then allow the appending and switching into an empty default block.

youknowriad commented:
> Not sure what you mean. Do you mean should the ↓ append a default block only when the block before it is a textual block?

Well! actually the opposite (I wrote the opposite of what I was thinking :) ). It should append a default block only when the block before is not a textual block because in a textual block (think a unique text area for everything), clicking ↓ does not create an empty line.

But I can be on board with the current behavior, just wanted to bring this up :)

youknowriad commented on editor/components/writing-flow/index.js:
```diff
@@ -98,6 +98,33 @@ class WritingFlow extends Component {
+		}
+
+		// Find block-level ancestor of said last tabbable
+		const blockEl = lastTabbable.closest( '.editor-block-list__block-edit' );
```
Should we use a const for the classname as it's being used in several places.

mcsf replied:
@youknowriad, what do you think of 31f9e0a35eb7a46b7749e12baf35779dfefa72b9? As an alternative to spinning up a `constants` file, etc.

youknowriad replied:
I love it

youknowriad reviewed (approved):
Left some optional comments but this is nice

mcsf committed:
BlockListBlock: Expose className constant

youknowriad commented on editor/components/block-list/block.js:
```diff
@@ -532,6 +532,8 @@ const mapDispatchToProps = ( dispatch, ownProps ) => ( {
 	},
 } );
 
+BlockListBlock.className = 'editor-block-list__block-edit';
```
❤️

mcsf commented:
> Looks like there's a visual regression in the blockquote citation. The font should be smaller and gray, which it is in master.

I see the regression, but I'm also seeing it in `master`. Can anyone confirm?

aduth commented on editor/effects.js:
```diff
@@ -388,4 +390,8 @@ export default {
 	},
+	APPEND_DEFAULT_BLOCK( action, store ) {
+		const { dispatch } = store;
+		dispatch( insertBlock( createBlock( getDefaultBlockName() ) ) );
```
Minor: Effects can simply `return` an action object to dispatch, if they're synchronous.

mcsf replied:
Ah, good tip, easy to forget.

aduth commented on editor/components/writing-flow/index.js:
```diff
@@ -98,6 +99,41 @@ class WritingFlow extends Component {
+
+		return blockDescendants.some( isElementNonEmpty );
+
+		function isElementNonEmpty( el ) {
```
Minor: Based on its name, I'd be inclined to expect `typeof isElementNonEmpty( el ) === 'boolean'`

aduth replied:
Hoisting mechanics are non-obvious 😬  Would it be sensible to move this to the top of the function, or out of it?

aduth commented on editor/components/writing-flow/index.js:
```diff
@@ -98,6 +99,41 @@ class WritingFlow extends Component {
+			.slice( blockIndex + 1 )
+			.filter( ( el ) => blockEl.contains( el ) );
+
+		return blockDescendants.some( isElementNonEmpty );
```
Minor: Seems an optimization could check both `blockEl.contains` and `isElementNonEmpty` in the same `Array#some`, instead of the separate `Array#filter`.

mcsf commented:
All good tips, @aduth. Amended.

mtias commented:
This is cool, nice work.
